### PR TITLE
Fix typographic substitution in (pre|code|kbd|script) blocks when smartypants=true

### LINKED ISSUE #1334: SmartyPants Shouldn't Do Typographic Substitutions In <code> Blocks
When SmartyPants is enabled, Marked is replacing `--` with an en-dash inside `<code>` blocks.

Per the following comment  on [the SmartyPants website](https://daringfireball.net/projects/smartypants/), I don't think it should.

> SmartyPants does not modify characters within `<pre>`, `<code>`, `<kbd>`, or `<script>` tag blocks.

**To Reproduce**

```bash
git clone https://github.com/richdouglasevans/marked-code-issue.git
cd marked-code-issue
npm i
npm test
```

> I am happy to dig into the code and sort this myself; I'm raising an issue so that I can first confirm with you the expected behaviour.
>
> Issue was noticed here: https://github.com/cypress-io/cypress-documentation/issues/923

### PR DESCRIPTION
**Pull request by richdouglasevans**
**Marked version:**

`0.5.0` (latest)

**Markdown flavor:** all

## Description

- Fixes #1334 

## Contributor

- [X] Test(s) exist to ensure functionality and minimize regression (if no tests added, list tests covering this PR); or,
- [ ] no tests required for this PR.
- [ ] If submitting new feature, it has been documented in the appropriate places.

## Committer

In most cases, this should be a different person than the contributor.

- [x] Draft GitHub release notes have been updated.
- [x] CI is green (no forced merge required).
- [x] Merge PR

### CODE REVIEW

File: lib/marked.js [hunk 1]
```diff
@@ -730,6 +730,14 @@ InlineLexer.prototype.output = function(src) {
       } else if (this.inLink && /^<\/a>/i.test(cap[0])) {
         this.inLink = false;
       }
+      if (!this.inRawBlock
+          && /^<pre|^<code|^<kbd|^<script/i.test(cap[0])) {
+        this.inRawBlock = true;
+      } else if (this.inRawBlock
+          && /^<\/pre>|^<\/code>|^<\/kbd>|^<\/script>/i.test(cap[0])) {
+        this.inRawBlock = false;
+      }
```
Comment (richdouglasevans) [inline lib/marked.js]: There's some duplication here: it'd be nice to just have a simple list of tags in which no substitution is to be performed -- `['pre', 'code', 'kbd', 'script']` -- 'cos that'd make it easy to add/remove tags; I did try that route but the regexes looked less "on the nose" than this version.

I also experimented with alternate names for the `inRawBlock` field: they were all too darn long and made the code less readable: forex, `isBlockInWhichNoSubstitutionIsToBePerformed`. Happy to take a steer here 😄

File: test/new/smartypants_code.md [hunk 2]
```diff
@@ -0,0 +1,9 @@
+---
+smartypants: true
+description: SmartyPants does not modify characters within <pre>, <code>, <kbd>, or <script> tag blocks.
+spec: https://daringfireball.net/projects/smartypants/
+---
```
Comment (richdouglasevans) [inline test/new/smartypants_code.md]: Just trying to be helpful by linking to the "spec" that explains _why_.

File: test/new/smartypants_code.md [hunk 3]
```diff
@@ -0,0 +1,9 @@
...
+<pre>&amp;</pre>
+<code>--foo</code>
+<kbd>---foo</kbd>
+<script>--foo</script>
```
Comment (richdouglasevans) [inline test/new/smartypants_code.md]: I was pleasantly surprised by the form the test suite takes: adding a specification was _so easy_, good work Team Marked.

😍

File: lib/marked.js (hunk 1 above)
Comment (UziTech) [reply lib/marked.js]: Could you change them to `/^<(pre|code|kbd|script)/i` and `/^<\/(pre|code|kbd|script)>/i`

I think that makes it a little easier to see what the regex is looking for

**Comment by richdouglasevans** [comment]
> I think that makes it a little easier to see what the regex is looking for.

Agreed, great suggestion 👍 I have done so.

File: lib/marked.js [hunk 4]
```diff
@@ -730,6 +730,14 @@ InlineLexer.prototype.output = function(src) {
       } else if (this.inLink && /^<\/a>/i.test(cap[0])) {
         this.inLink = false;
       }
+      if (!this.inRawBlock
+          && /^<(pre|code|kbd|script)/i.test(cap[0])) {
```
Comment (UziTech) [inline lib/marked.js]: There should be a `(\s|>)` at the end to prevent matching other tags that start with these words

Comment (richdouglasevans) [reply lib/marked.js]: Good one Tony; I've done so and expanded the tests to prove that.

> Loving your attention to detail here, much appreciated 👍

**Comment by styfle** [comment]
Great job! 🎉 

coach z

**Pull request merged**